src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,183 @@
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Threading.Channels;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+internal sealed class WebTransportSession : IWebTransportSession
+{
+    private static readonly IStreamDirectionFeature _outputStreamDirectionFeature = new DefaultStreamDirectionFeature(canRead: false, canWrite: true);
+
+    private readonly CancellationTokenRegistration _connectionClosedRegistration;
+
+    // stores all created streams (pending or accepted)
+    private readonly ConcurrentDictionary<long, WebTransportStream> _openStreams = new();
+    // stores all pending streams that have not been accepted yet
+    private readonly Channel<WebTransportStream> _pendingStreams;
+
+    private readonly Http3Connection _connection;
+    private readonly Http3Stream _connectStream = default!;
+    private bool _isClosing;
+
+    private static readonly ReadOnlyMemory<byte> OutputStreamHeader = new(new byte[] {
+            0x40 /*quic variable-length integer length*/,
+            (byte)Http3StreamType.WebTransportUnidirectional,
+            0x00 /*body*/});
+
+    internal static string WebTransportProtocolValue => "webtransport";
+    internal static string SecPrefix => "sec-webtransport-http3-";
+    internal static string VersionHeaderPrefix => $"{SecPrefix}draft";
+    internal static string CurrentSuppportedVersion => $"{VersionHeaderPrefix}02";
+
+    long IWebTransportSession.SessionId => _connectStream.StreamId;
+
+    internal WebTransportSession(Http3Connection connection, Http3Stream connectStream)
+    {
+        _connection = connection;
+        _connectStream = connectStream;
+        _isClosing = false;
+        // unbounded as limits to number of streams is enforced elsewhere
+        _pendingStreams = Channel.CreateUnbounded<WebTransportStream>();
+
+        // listener to abort if this connection is closed
+        _connectionClosedRegistration = connection._multiplexedContext.ConnectionClosed.Register(static state =>
+        {
+            var session = (WebTransportSession)state!;
+            session.OnClientConnectionClosed();
+        }, this);
+    }
+
+    void IWebTransportSession.Abort(int errorCode)
+    {
+        Abort(new(), (Http3ErrorCode)errorCode);
+    }
+
+    internal void OnClientConnectionClosed()
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            foreach (var stream in _openStreams)
+            {
+                _ = stream.Value.DisposeAsync().AsTask();
```

Comment:
Red flag: you should almost never discard a task. Is there a non-async Dispose? Otherwise block.

`.GetAwaiter().GetResult();`